### PR TITLE
Avoid `prepare_metadata_for_build_wheel` calls for Hatch packages with dynamic dependencies

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Hatch allows for highly dynamic customization of metadata via hooks. In such cases, Hatch
can't upload the PEP 517 contract, in that the metadata Hatch would return by
`prepare_metadata_for_build_wheel` isn't guaranteed to match that of the built wheel.

Hatch disables `prepare_metadata_for_build_wheel` entirely for pip. We'll instead disable
it on our end when metadata is defined as "dynamic" in the pyproject.toml, which should
allow us to leverage the hook in _most_ cases while still avoiding incorrect metadata for
the remaining cases.

Closes: https://github.com/astral-sh/uv/issues/2130.

### CODE REVIEW

**Comment by charliermarsh** [comment]
Just the failing test to start.

File: scripts/packages/hatch_dynamic/requirements-cli.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+black
```
Comment (ofek) [inline scripts/packages/hatch_dynamic/requirements-cli.txt:1]: This file can be deleted now as well as the other requirements file

File: crates/uv-build/src/lib.rs [hunk 2]
```diff
@@ -607,6 +616,31 @@ impl SourceBuild {
             return Ok(Some(metadata_dir.clone()));
         }
 
+        // Hatch allows for highly dynamic customization of metadata via hooks. In such cases, Hatch
+        // can't upload the PEP 517 contract, in that the metadata Hatch would return by
```
Comment (zanieb) [inline crates/uv-build/src/lib.rs:620]:
```suggestion
        // can't uphold the PEP 517 contract, in that the metadata Hatch would return by
```

**Review by zanieb — APPROVED** [review]

File: crates/uv-build/src/lib.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/uv-build/src/lib.rs:620]: Upload the contract

**Comment by potiuk** [comment]
Installed `uv` from main and I confirm it fixes the issue. Both - local installation from "file://" and remote via URL work.

**Pull request merged**